olp-cpp-sdk-core/src/client/OlpClient.cpp

```diff
@@ -240,31 +240,44 @@ std::shared_ptr<http::NetworkRequest> OlpClient::CreateRequest(
 }
 
 NetworkAsyncCallback GetRetryCallback(
-    int current_try, int current_backdown_period, const RetrySettings& settings,
+    int current_try, int current_backdown_period,
+    long long accumulated_wait_time, const RetrySettings& settings,
     const NetworkAsyncCallback& callback,
     const std::shared_ptr<http::NetworkRequest>& network_request,
     std::weak_ptr<http::Network> network,
     const std::weak_ptr<CancellationContext>& weak_cancel_context) {
   ++current_try;
   return [=](HttpResponse response) {
+    const long long max_wait_time =
+        settings.timeout * 1000ll;  // convert seconds to msecs
     if (current_try > settings.max_attempts ||
-        !settings.retry_condition(response)) {
+        !settings.retry_condition(response) ||
+        accumulated_wait_time >= max_wait_time) {
       callback(std::move(response));
     } else {
       // TODO there must be a better way
-      std::this_thread::sleep_for(
-          std::chrono::milliseconds(current_backdown_period));
-
-      auto next_backdown_period =
-          settings.backdown_policy(current_backdown_period);
+      // Make sure that we don't wait longer than `timeout` in retry settings
+      const long long actual_wait_time = std::min<long long>(
```

Comment:
```suggestion
      const std::chrono::milliseconds actual_wait_time = std::chrono::milliseconds::min(
```
Prefer abstract types over built-in types.